trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/rest/RequestFilter.java

```diff
@@ -0,0 +1,204 @@
+    {
+        private InternalRequestContext
+        {
+            requireNonNull(request, "request is null");
+            requireNonNull(signingMetadata, "signingMetadata is null");
+            requireNonNull(requestLoggingSession, "requestLoggingSession is null");
+        }
+    }
+
+    @Inject
+    RequestFilter(SigningController signingController, Map<Class<?>, SigningServiceType> signingServiceTypesMap, RequestLoggerController requestLoggerController)
+    {
+        this.signingController = requireNonNull(signingController, "signingController is null");
+        this.signingServiceTypesMap = ImmutableMap.copyOf(signingServiceTypesMap);
+        this.requestLoggerController = requireNonNull(requestLoggerController, "requestLoggerController is null");
+    }
+
+    @SuppressWarnings("ThrowableNotThrown")
+    @Override
+    public void filter(ContainerRequestContext requestContext)
+            throws IOException
+    {
+        if (requestContext.getRequest() instanceof ContainerRequest containerRequest) {
+            if (containerRequest.getUriInfo().getMatchedResourceMethod() == null) {
+                log.warn("%s does not have a MatchedResourceMethod", containerRequest.getRequestUri());
+                throw new WebApplicationException(INTERNAL_SERVER_ERROR);
+            }
+
+            Class<?> declaringClass = containerRequest.getUriInfo().getMatchedResourceMethod().getInvocable().getDefinitionMethod().getDeclaringClass();
+            SigningServiceType signingServiceType = signingServiceTypesMap.get(declaringClass);
+            if (signingServiceType == null) {
+                log.warn("%s does not have a SigningServiceType", declaringClass.getName());
+                throw new WebApplicationException(INTERNAL_SERVER_ERROR);
+            }
+
+            Request request = RequestBuilder.fromRequest(containerRequest);
+            RequestLoggingSession requestLoggingSession = requestLoggerController.newRequestSession(request, signingServiceType);
+            SigningMetadata signingMetadata;
+            try {
+                signingMetadata = signingController.validateAndParseAuthorization(request, signingServiceType);
+            }
+            catch (Exception e) {
+                requestLoggingSession.logException(e);
+
+                try {
+                    switch (Throwables.getRootCause(e)) {
+                        case WebApplicationException webApplicationException -> throw webApplicationException;
+                        case IOException ioException -> throw ioException;
+                        default -> throw new RuntimeException(e);
+                    }
+                }
+                finally {
+                    requestLoggingSession.close();
+                }
+            }
+
+            containerRequest.setProperty(RequestFilter.class.getName(), new InternalRequestContext(request, signingMetadata, requestLoggingSession));
```

Comment:
Can we add each class as a property without wrapping them with `InternalRequestContext`?

The current issue is we can't "bind" `RequestLoggingSession` because we aren't done parsing the request, so we have more closing logic (the `finally` block above).

If we add each class on its own with:
```java
containerRequest.setProperty(Request.class.getName(), request);
```
(for each class);

We can then update the `unwarpContext` function to:

```java
    private static <T> T unwrapContext(ContainerRequest containerRequest, Class<? extends T> clazz)
    {
        return clazz.cast(containerRequest.getProperty(clazz.getName()));
    }
```

And in the response filter we get just the `RequestLoggingSession` and close it.

WDYT?

Reply:
Good idea. done